File: projects/06/assembler/codegen.py
from typing import Optional


class Code:
    """
    Translates Hack assembly language mnemonics into binary codes.
    """

    def __init__(self):
        pass

    def dest(self, mnemonic: Optional[str]) -> str:
        """
        Returns the binary code of the dest mnemonic
        """
        if mnemonic is None:
            result = "000"
        elif mnemonic == "M":
            result = "001"
        elif mnemonic == "D":
            result = "010"
        elif mnemonic == "MD":
            result = "011"
        elif mnemonic == "A":
            result = "100"
        elif mnemonic == "AM":
            result = "101"
        elif mnemonic == "AD":
            result = "110"
        else:
            result = "111"
        return result

    def comp(self, mnemonic: str) -> str:
        """
        Returns the binary code of the comp mnemonic
        """
        if mnemonic == "0":
            result = "0101010"
        elif mnemonic == "1":
            result = "0111111"
        elif mnemonic == "-1":
            result = "0111010"
        elif mnemonic == "D":
            result = "0001100"
        elif mnemonic == "A":
            result = "0110000"
        elif mnemonic == "M":
            result = "1110000"
        elif mnemonic == "!D":
            result = "0001101"
        elif mnemonic == "!A":
            result = "0110001"
        elif mnemonic == "!M":
            result = "1110001"
        elif mnemonic == "-D":
            result = "0001111"
        elif mnemonic == "-A":
            result = "0110011"
        elif mnemonic == "-M":
            result = "1110011"
        elif mnemonic == "D+1":
            result = "0011111"
        elif mnemonic == "A+1":
            result = "0110111"
        elif mnemonic == "M+1":
            result = "1110111"
        elif mnemonic == "D-1":
            result = "0001110"
        elif mnemonic == "A-1":
            result = "0110011"
        elif mnemonic == "M-1":
            result = "1110011"
        elif mnemonic == "D+A":
            result = "0000110"
        elif mnemonic == "D+M":
            result = "1000110"
        elif mnemonic == "D-A":
            result = "0010011"
        elif mnemonic == "D-M":
            result = "1010011"
        elif mnemonic == "A-D":
            result = "0000111"
        elif mnemonic == "M-D":
            result = "1000111"
        elif mnemonic == "D&A":
            result = "0000000"
        elif mnemonic == "D&M":
            result = "1000000"
        elif mnemonic == "D|A":
            result = "0010101"
        else:
            result = "1010101"
        return result

    def jump(self, mnemonic: Optional[str]) -> str:
        """
        Returns the binary code of the jump mnemonic
        """
        if mnemonic is None:
            result = "000"
        elif mnemonic == "JGT":
            result = "001"
        elif mnemonic == "JEQ":
            result = "010"
        elif mnemonic == "JGE":
            result = "011"
        elif mnemonic == "JLT":
            result = "100"
        elif mnemonic == "JNE":
            result = "101"
        elif mnemonic == "JLE":
            result = "110"
        else:
            result = "111"
        return result


def test_jump():
    """
    Test for the jump method.
    """
    code_gen = Code()
    assert code_gen.jump(None) == "000"
    assert code_gen.jump("JGT") == "001"
    assert code_gen.jump("JEQ") == "010"
    assert code_gen.jump("JGE") == "011"
    assert code_gen.jump("JLT") == "100"
    assert code_gen.jump("JNE") == "101"
    assert code_gen.jump("JLE") == "110"
    assert code_gen.jump("JMP") == "111"


def test_dest():
    """
    Test for dest method.
    """
    code_gen = Code()
    assert code_gen.dest(None) == "000"
    assert code_gen.dest("M") == "001"
    assert code_gen.dest("D") == "010"
    assert code_gen.dest("MD") == "011"
    assert code_gen.dest("A") == "100"
    assert code_gen.dest("AM") == "101"
    assert code_gen.dest("AD") == "110"
    assert code_gen.dest("AMD") == "111"


def test_comp():
    pass

File: projects/06/rasm/src/code_gen.rs
// Translates Hack assembly language mnemonics into binary codes.
#[allow(dead_code)]
pub(crate) struct Code;

#[allow(dead_code)]
impl Code {
    // Returns the binary code of the dest mnemonic

    pub(crate) fn dest(&self, mnemonic: Option<&str>) -> &'static str {
        match mnemonic {
            Some(mne) => match mne {
                "M" => "001",
                "D" => "010",
                "MD" => "011",
                "A" => "100",
                "AM" => "101",
                "AD" => "110",
                _ => "111",
            },
            _ => "000",
        }
    }

    // Returns the binary code of the comp mnemonic
    pub(crate) fn comp(&self, mnemonic: &str) -> &'static str {
        match mnemonic {
            "0" => "0101010",
            "1" => "0111111",
            "-1" => "0111010",
            "D" => "0001100",
            "A" => "0110000",
            "M" => "1110000",
            "!D" => "0001101",
            "!A" => "0110001",
            "!M" => "1110001",
            "-D" => "0001111",
            "-A" => "0110011",
            "-M" => "1110011",
            "D+1" => "0011111",
            "A+1" => "0110111",
            "M+1" => "1110111",
            "D-1" => "0001110",
            "A-1" => "0110010",
            "M-1" => "1110010",
            "D+A" => "0000010",
            "D+M" => "1000010",
            "D-A" => "0010011",
            "D-M" => "1010011",
            "A-D" => "0000111",
            "M-D" => "1000111",
            "D&A" => "0000000",
            "D&M" => "1000000",
            "D|A" => "0010101",
            _ => "1010101",
        }
    }

    // Returns the binary code of the jump mnemonic
    pub(crate) fn jump(&self, mnemonic: Option<&str>) -> &'static str {
        match mnemonic {
            Some(mne) => match mne {
                "JGT" => "001",
                "JEQ" => "010",
                "JGE" => "011",
                "JLT" => "100",
                "JNE" => "101",
                "JLE" => "110",
                _ => "111",
            },
            _ => "000",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_jump() {
        let c = Code {};
        assert_eq!(c.jump(Some("JLT")), "100")
    }

    #[test]
    fn test_dest() {
        let c = Code {};
        assert_eq!(c.dest(Some("M")), "001")
    }

    #[test]
    fn test_comp() {
        let c = Code {};
        assert_eq!(c.comp("M-D"), "1000111");
    }
}
